#include<stdio.h>

void usage_phm(){
	fprintf(stderr, "phm:\n");
	fprintf(stderr, "  Pairwise Hierarchical Model\n");
	fprintf(stderr, "  A Bayesian hierarchical model to map causal interactions between regulatory elements in the genome, that incorporates techniques from Mendelian Randomisation\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "Usage:\n");
	fprintf(stderr, "  phm -i <input file> -c <col type string> -r <N rows> -f <N features> -o <output dir> [OPTIONS]\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "Input:\n");
	fprintf(stderr, "  Output of bayesian regression which consists of\n");
	fprintf(stderr, "  paired peak IDs with Regional Bayes factors for 10 hypotheses.\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "Output:\n");
	fprintf(stderr, "  Posterior probability and coefficients for covariates.\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "Mandatory arguments:\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "  -i/--input                  <string>  Input file\n");
	fprintf(stderr, "  -r/--n-rows                 <int>     Number of rows in input\n");
	fprintf(stderr, "  -f/--n-features             <int>     Number of features\n");
	fprintf(stderr, "  -c/--col-type               <string>  Column type string (e.g., J,K,N4,B10)\n");
	fprintf(stderr, "  -o/--output                 <string>  Output directory name\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "Other options:\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "  -t/--n-threads              <int>     The number of threads used in the E-step\n");
	fprintf(stderr, "  -v/--verbose                          Print verbose output (for debugging).\n");
	fprintf(stderr, "  \n");
}

void usage_hm(){
	fprintf(stderr, "hm:\n");
	fprintf(stderr, "  Hierarchical Model\n");
	fprintf(stderr, "  Introduced by Veyrieras et al. (2008)\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "Usage:\n");
	fprintf(stderr, "  hm -i <input file> -c <col type string> -r <N rows> -f <N features> -o <output dir> [OPTIONS]\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "Input:\n");
	fprintf(stderr, "  Association Bayes factors and various annotations\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "Output:\n");
	fprintf(stderr, "  Hyperparameter estimate for variant-level and feature-level prior probabilities \n");
	fprintf(stderr, "  \n");
	fprintf(stderr, "Mandatory arguments:\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "  -i/--input                  <string>  Input file\n");
	fprintf(stderr, "  -r/--n-rows                 <int>     Number of rows in input\n");
	fprintf(stderr, "  -f/--n-features             <int>     Number of features\n");
	fprintf(stderr, "  -c/--col-type               <string>  Column type string (e.g., I,C3,C2,B,N0,N4)\n");
	fprintf(stderr, "  -o/--output                 <string>  Output directory name\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "Feature level annotation options:\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "  -j/--feature-input          <string>  Covariates of feature-level prior probability\n");
	fprintf(stderr, "  -d/--col-type-feature       <string>  Column type of feature covariates\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "Output options:\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "  -p/--posterior-probability            Output posterior probabilities\n");
	fprintf(stderr, "  --credible-set                        Output the numbers of variants in the credible sets (90%, 95% and 99%)\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "Other options:\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "  -t/--n-threads              <int>     The number of threads used in the E-step\n");
	fprintf(stderr, "  -v/--verbose                          Print verbose output (for debugging). \n");
	fprintf(stderr, "  --init-variant-level        <string>  Initial parameter values for the variant-level prior probability\n");
	fprintf(stderr, "\n");
}

void usage_bayeslm(){
	fprintf(stderr, "bayeslm:\n");
	fprintf(stderr, "  Bayesian linear regression model\n");
	fprintf(stderr, "  Computing asymptotic Bayes factors (BFs) and regional Bayes factors (RBFs)\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "Usage:\n");
	fprintf(stderr, "  bayeslm -g <vcf file> -i <fragment count> -j <peak ID> -p <bed file> \\\n");
	fprintf(stderr, "          -s <chromosome> -c <position> -w <width> [OPTIONS]\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "Input:\n");
	fprintf(stderr, "  VCF file / Normalized fragment counts / peak annotation in BED format\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "Output:\n");
	fprintf(stderr, "  Bayes factors and variant-level annotations.\n");
	fprintf(stderr, "  \n");
	fprintf(stderr, "Mandatory arguments:\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "  -g/--vcf                   <string>  VCF file \n");
	fprintf(stderr, "  -i/--normalised-count      <string>  Normalized fragment counts in binary double (e.g., FPKM, TPM, etc.)\n");
	fprintf(stderr, "  -j/--feature-id            <int>     Feature ID (row number of the BED file specified below)\n");
	fprintf(stderr, "  -p/--feature-bed           <string>  Peak annotation in BED format (must be tabix indexed)\n");
	fprintf(stderr, "  -s/--window-chromosome     <int>     Coordinate at which cis-window is centred\n");
	fprintf(stderr, "  -c/--window-centre         <int>     Coordinate at which cis-window is centred\n");
	fprintf(stderr, "  -w/--window-size           <int>     cis-window size (default: 1000000)\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "Colocalisation options:\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "  -g2/--vcf2                 <string>  VCF file for the second trait\n");
	fprintf(stderr, "  -i2/--normalised-count2    <string>  Normalized fragment counts for the second trait\n");
	fprintf(stderr, "  -j2/--feature-id2          <int>     Feture ID for the second trait\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "Prior probability options (used in RBF and lead variant calculation):\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "  -p1/--variant-level        <string>  Hyperparameters of variant-level prior probability\n");
	fprintf(stderr, "  -p2/--feature-level        <string>  Hyperparameters of peak-level prior probability\n");
	fprintf(stderr, "  -p3/--feature-pair-level   <string>  Hyperparameters of peak-level prior probability\n");
	fprintf(stderr, "  --variant-level2           <string>  Hyperparameters of variant-level prior probability for the second trait\n");
	fprintf(stderr, "  --feature-level2           <string>  Hyperparameters of peak-level prior probability for the second trait\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "Output options:\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "  -o/--output                <string>  Output file name\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "Other options:\n");
	fprintf(stderr, "\n");
	fprintf(stderr, "  -v/--verbose                         Print verbose output in stderr (for debugging).\n");
	fprintf(stderr, "  -r/--random-permutation              Permute samples\n");
}
